pkg/dataobj/consumer/partition_processor.go

```diff
@@ -107,13 +114,39 @@ func (p *partitionProcessor) start() {
 
 func (p *partitionProcessor) stop() {
 	p.cancel()
+	p.runningLock.Lock()
+	p.finished = true
+	close(p.records)
+	p.runningLock.Unlock()
 	p.wg.Wait()
 	if p.builder != nil {
 		p.builder.UnregisterMetrics(p.reg)
 	}
 	p.metrics.unregister(p.reg)
 }
 
+// Drops records from the channel if the processor is stopped.
+// Returns false if the processor is stopped, true otherwise.
+func (p *partitionProcessor) Append(records []*kgo.Record) bool {
```

Comment:
small tip: Go channels do not _have_ to be closed when you're done with them; they can still be garbage collected as long as they're not being held anywhere from any Go routine. 

Relying on that could help clean some of this logic up and help remove the dependency on the extra mutex/finished flag.